src/plugins/discover/public/application/components/data_visualizer_grid/data_visualizer_grid.tsx

```diff
@@ -0,0 +1,169 @@
+  indexPattern: DataView;
+  savedSearch?: SavedSearch;
+  query?: Query;
+  visibleFieldNames?: string[];
+  filters?: Filter[];
+  showPreviewByDefault?: boolean;
+}
+export interface DataVisualizerGridEmbeddableOutput extends EmbeddableOutput {
+  showDistributions?: boolean;
+}
+
+export interface DiscoverDataVisualizerGridProps {
+  /**
+   * Determines which columns are displayed
+   */
+  columns: string[];
+  /**
+   * The used data view
+   */
+  indexPattern: DataView;
+  /**
+   * Saved search description
+   */
+  searchDescription?: string;
+  /**
+   * Saved search title
+   */
+  searchTitle?: string;
+  /**
+   * Discover plugin services
+   */
+  services: DiscoverServices;
+  savedSearch?: SavedSearch;
+  query?: Query;
+  filters?: Filter[];
+  stateContainer?: GetStateReturn;
+}
+
+export const EuiDataGridMemoized = React.memo((props: EuiDataGridProps) => {
+  return <EuiDataGrid {...props} />;
+});
+
+export const DiscoverDataVisualizerGrid = (props: DiscoverDataVisualizerGridProps) => {
+  const { services, indexPattern, savedSearch, query, columns, filters, stateContainer } = props;
+  const { uiSettings } = services;
+
+  const [embeddable, setEmbeddable] = useState<
+    | ErrorEmbeddable
+    | IEmbeddable<DataVisualizerGridEmbeddableInput, DataVisualizerGridEmbeddableOutput>
+    | undefined
+  >();
+  const embeddableRoot: React.RefObject<HTMLDivElement> = useRef<HTMLDivElement>(null);
+
+  const showPreviewByDefault = useMemo(
+    () =>
+      stateContainer ? !stateContainer.appStateContainer.getState().hideAggregatedPreview : true,
+    [stateContainer]
+  );
+
+  useEffect(() => {
+    embeddable?.getOutput$().subscribe((output: DataVisualizerGridEmbeddableOutput) => {
+      if (output.showDistributions !== undefined && stateContainer) {
+        stateContainer.setAppState({ hideAggregatedPreview: !output.showDistributions });
+      }
+    });
+  }, [embeddable, stateContainer]);
+
+  useEffect(() => {
+    if (embeddable && !isErrorEmbeddable(embeddable)) {
+      // Update embeddable whenever one of the important input changes
+      embeddable.updateInput({
+        indexPattern,
+        savedSearch,
+        query,
+        filters,
+        visibleFieldNames: columns,
+      });
+      embeddable.reload();
+    }
+  }, [embeddable, indexPattern, savedSearch, query, columns, filters]);
+
+  useEffect(() => {
+    if (showPreviewByDefault && embeddable && !isErrorEmbeddable(embeddable)) {
+      // Update embeddable whenever one of the important input changes
+      embeddable.updateInput({
+        showPreviewByDefault,
+      });
+      embeddable.reload();
+    }
+  }, [showPreviewByDefault, uiSettings, embeddable]);
+
+  useEffect(() => {
+    return () => {
+      // Clean up embeddable upon unmounting
+      embeddable?.destroy();
+    };
+  }, [embeddable]);
+
+  useEffect(() => {
+    let unmounted = false;
+    const loadEmbeddable = async () => {
+      if (services?.embeddable) {
```

Comment:
Is there a way services can be undefined?

Reply:
Good call! Removed the checks here https://github.com/elastic/kibana/pull/107184/commits/ba1cad7ff6ee7b3391fb42000306c6b3b96096d5